### PR TITLE
Fix the "tensors to be on the same device" error in HistogramObserver

### LINKED ISSUE #59075: Quantization-aware training: HistogramObserver raises "tensors to be on the same device" error.
## 🐛 Bug
The `HistogramObserver` raises `RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu!` error when perform quantization-aware training on the GPU.

The reason is that the histogram is calculated through input activation `x` which is on the GPU:
https://github.com/pytorch/pytorch/blob/56b43f4fec1f76953f15a627694d4bba34588969/torch/quantization/observer.py#L916

while other arguments to the `_get_norm()` function is on the CPU:
https://github.com/pytorch/pytorch/blob/56b43f4fec1f76953f15a627694d4bba34588969/torch/quantization/observer.py#L771-L775

I think a simple fix would be transporting the histogram to CPU manually, or alternatively transfering other tensors
via` to(self.histogram)`.

The `MovingAverageMinMaxObserver` observer works fine for me. The error occurs only when the `HistogramObserver` is used.

## To Reproduce
Code to reproduce the error:
```
import torch, torchvision
import torch.quantization as quant
qconfig = quant.QConfig(
    activation=quant.FakeQuantize.with_args(
        observer=quant.fake_quantize.HistogramObserver,
        quant_min=0,
        quant_max=255,
        dtype=torch.quint8,
        qscheme=torch.per_tensor_symmetric,
        reduce_range=False),
    weight=quant.FakeQuantize.with_args(
        observer=quant.fake_quantize.MovingAverageMinMaxObserver,
        quant_min=-128,
        quant_max=127,
        dtype=torch.qint8,
        qscheme=torch.per_tensor_symmetric,
        reduce_range=False))
model = torchvision.models.quantization.mobilenet_v2()
model.qconfig = qconfig
quant.prepare_qat(model, inplace=True)
model.cuda()
model(torch.rand(12, 3, 224, 224).cuda())
```

## Environment
```
PyTorch version: 1.8.1+cu111
Is debug build: False
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: Tencent tlinux 2.2 (Final) (x86_64)
GCC version: (GCC) 4.8.5 20150623 (Red Hat 4.8.5-4)
Clang version: Could not collect
CMake version: version 3.12.3
Libc version: glibc-2.17

Python version: 3.8 (64-bit runtime)
Python platform: Linux-3.10.107-1-tlinux2_kvm_guest-0049-x86_64-with-glibc2.10
Is CUDA available: True
CUDA runtime version: 11.0.194
GPU models and configuration: GPU 0: GRID T4-8C
Nvidia driver version: 450.102.04
cuDNN version: Probably one of the following:
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn.so.7
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn.so.8
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_adv_train.so.8
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8
/usr/local/cuda-11.0/targets/x86_64-linux/lib/libcudnn_ops_train.so.8
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.2
[pip3] torch==1.8.1+cu111
[pip3] torchvision==0.9.1
[conda] numpy                     1.19.2                   pypi_0    pypi
[conda] torch                     1.8.1+cu111              pypi_0    pypi
[conda] torchvision               0.9.1                    pypi_0    pypi
```

## Additional context
Strangely, use the default qconfig `quant.get_default_qconfig("qnnpack")` will not incur this error though both
the qconfigs use the `HistogramObserver`.

The following code runs smoothly without an error.
```
import torch, torchvision
import torch.quantization as quant

model = torchvision.models.quantization.mobilenet_v2()
model.qconfig = quant.get_default_qconfig("qnnpack")
model.qconfig = qconfig
print(model.qconfig)
quant.prepare_qat(model, inplace=True)

model.cuda()
model(torch.rand(12, 3, 224, 224).cuda())
```

The details of `quant.get_default_qconfig("qnnpack")` is
```
QConfig(activation=functools.partial(<class 'torch.quantization.observer.HistogramObserver'>, reduce_range=False), weight=functools.partial(<class 'torch.quantization.observer.MinMaxObserver'>, dtype=torch.qint8, qscheme=torch.per_tensor_symmetric))
```

cc @jerryzh168 @jianyuh @raghuramank100 @jamesr66a @vkuzo

### PR DESCRIPTION
**Pull request by zeakey**
Fixes #59075

This PR fixes the "tensors to be on the same device" error in `HistogramObserver`.

### CODE REVIEW

File: torch/quantization/observer.py [hunk 1]
```diff
@@ -878,7 +878,7 @@ def _compute_quantization_error(self, next_start_bin: int, next_end_bin: int):
         if dst_bin_width == 0.0:
             return 0.0
 
-        src_bin = torch.arange(self.bins)
+        src_bin = torch.arange(self.bins).to(self.histogram.device)
```
Comment (ngimel) [inline torch/quantization/observer.py:881]:
```suggestion
        src_bin = torch.arange(self.bins, device=self.histogram.device)
```

File: torch/quantization/observer.py [hunk 2]
```diff
@@ -897,11 +897,13 @@ def _compute_quantization_error(self, next_start_bin: int, next_end_bin: int):
 
         density = self.histogram / bin_width
 
-        norm = torch.zeros(self.bins)
+        norm = torch.zeros(self.bins).to(self.histogram.device)
```
Comment (vkuzo) [inline torch/quantization/observer.py:900]: thanks for fixing this!

could we make this

```
norm = torch.zeros(self.bins, device=self.histogram.device)
```

File: torch/quantization/observer.py [hunk 3]
```diff
@@ -897,11 +897,13 @@ def _compute_quantization_error(self, next_start_bin: int, next_end_bin: int):
...
 
         delta_begin = src_bin_begin - dst_bin_of_begin_center
         delta_end = dst_bin_width / 2
-        norm += self._get_norm(delta_begin, torch.ones(self.bins) * delta_end, density)
+        norm += self._get_norm(delta_begin,
+                               torch.ones(self.bins).to(self.histogram.device) * delta_end,
```
Comment (ngimel) [inline torch/quantization/observer.py:905]:
```suggestion
                               torch.ones(self.bins, device=self.histogram.device) * delta_end,
```

File: torch/quantization/observer.py (hunk 2 above)
Comment (ngimel) [inline torch/quantization/observer.py:900]:
```suggestion
        norm = torch.zeros(self.bins, device=self.histogram.device)
```

**Comment by ngimel** [comment]
@vkuzo I've fixed `device=` usage, can you please land if everything else is fine?

**Review by vkuzo — APPROVED** [review]

**Comment by vkuzo** [comment]
any chance we could rebase on a recent revision? I see some errors in OSS CI which look transient

**Pull request closed without merging**